### ISSUE TITLE
Build ultimate viewing station interface

### ISSUE THREAD

**Issue by micahflee** [freedomofpress/securedrop#415]
I think we should build a GUI program to be installed in the Tails persistent volume that holds the journalist key. Basically, my hope is to move most of the work that journalists do from their internet-connected Tails to their air-gapped viewing station Tails. The journalist workflow can change to this:
- On internet-connected Tails, log into SecureDrop
- Download latest changes file, something like `net-to-airgap.securedrop`
- Boot up viewing station Tails, and launch SecureDrop app
- Import `net-to-airgap.securedrop` in app, which will load all of the latest changes
- Use SecureDrop app to look at stuff, write replies (or start writing them and save drafts), star things, search things, etc.
- Export changes to something like `airgap-to-net.securedrop`
- Back on internet Tails, upload `airgap-to-net.securedrop` to post all replies to source

If we do this, we get some awesome benefits:
- Everything can be automatically organized. Right now the journalist has to do a massive amount of manual organization, creating intricate file structures, etc.
- Responses to sources get encrypted on the air-gapped computer, rather than on the connected-to-internet SecureDrop server.
- Journalists can keep track of what they said in response too. It's hard to follow a one-sided conversation, so this will be helpful.
- Journalists can keep detailed notes about sources, rename them to their actual names or whatever they want, etc., and it's not a problem because it's all getting saved on the air-gapped computer, never touching a computer with internet access.
- Journalists can have a usable interface to have per-document discussions (#391).
- If we do it right it will be _so much_ simpler for the journalist to use.

I recently have been playing with pywebkitgtk to build a [GUI for onionshare](https://github.com/micahflee/onionshare/tree/master/onionshare_gui) using HTML, CSS, and javascript for the GUI, but still being a standalone python program. Basically, you create a GTK window, put a webkit webview in it, and then render an HTML page. [This guide](http://www.aclevername.com/articles/python-webgui/) was really helpful.

I think by building it as a web app that runs as a program we will have a lot of flexibility, and can make it work really well and look really polished. Optionally we can also run a web server with flask there to serve the websites and use a sqlite database (in onionshare it just serves a straight html file with no web server).

In terms of downloading changes on the internet computer to move to the air-gap viewing station and visa-versa, we obviously need to only download diffs and not everything each time. There are various ways we could go about doing this, but I think we can think of the current state of a securedrop instance sort of like a git repo, and commits would happen each time a source is created, a doc or message is submitted, a reply is sent, or something is deleted. Each of these events happens chronologically, and they can have unique ids.

The securedrop server can keep track of which "commit" the journalist's air-gap is currently on. When they click the download changes button, it can bundle all "commits" from when they last downloaded until the latest commit, and serve them all together. And a similar thing can happen in the other direction, when the journalist exports from the air-gapped GUI app. If there are multiple journalists accessing the same securedrop, merges can be handled automatically. This stuff can all happen under the hood -- the only thing journalists should need to know is that they need to copy latest changes to their air-gapped computer, do some work, then copy latest changes back to their internet computer.

I think this will be a huge long-term project, like maybe part of the 1.0 roadmap. Thoughts?

**Comment by redshiftzero** [comment]
this is now being actively worked on in the qt client repo: https://github.com/freedomofpress/securedrop-client
